### PR TITLE
Unify determination of whether a volume is ephemeral

### PR DESCRIPTION
**Pull request by mrkm4ntr**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Determinations of whether a volume is ephemeral are duplicated and different from each other.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/volume/util/util.go [hunk 1]
```diff
@@ -545,6 +545,6 @@ func GetPluginMountDir(host volume.VolumeHost, name string) string {
 // volume vs. some other type
 func IsLocalEphemeralVolume(volume v1.Volume) bool {
 	return volume.GitRepo != nil ||
-		(volume.EmptyDir != nil && volume.EmptyDir.Medium != v1.StorageMediumMemory) ||
+		(volume.EmptyDir != nil && volume.EmptyDir.Medium == v1.StorageMediumDefault) ||
```
Comment (mattjmcnaughton) [inline pkg/volume/util/util.go]: Can we assume that `volume.EmptyDir.Medium != v1.StorageMediumMemory` is equivalent to `volume.EmptyDir.Medium == v1.StorageMediumDefault`?

File: pkg/kubelet/server/stats/volume_stat_calculator.go [hunk 2]
```diff
@@ -123,7 +124,7 @@ func (s *volumeStatCalculator) calcAndStoreStats() {
 			}
 		}
 		volumeStats := s.parsePodVolumeStats(name, pvcRef, metric, volSpec)
-		if isVolumeEphemeral(volSpec) {
+		if util.IsLocalEphemeralVolume(volSpec) {
```
Comment (mattjmcnaughton) [inline pkg/kubelet/server/stats/volume_stat_calculator.go]: At a high level, is there a difference between an "ephemeral" volume and a "local ephemeral" volume? Or is this just a naming distinction?

File: pkg/volume/util/util.go [hunk 3]
```diff
@@ -545,6 +545,6 @@ func GetPluginMountDir(host volume.VolumeHost, name string) string {
...
 		volume.ConfigMap != nil || volume.DownwardAPI != nil
```
Comment (mattjmcnaughton) [inline pkg/volume/util/util.go]: It seems like now some checks that didn't have `volume.DownwardAPI != nil` will now have that check. Could that have adverse impacts? What exactly is `volume.DownwardAPI` and when could it be nil?

**Review by mattjmcnaughton — COMMENTED** [review]
Thanks for this change :)

Overall, I support the goal of code unification. I made two comments just confirming that certain behaviors are equivalent.

File: pkg/volume/util/util.go (hunk 1 above)
Comment (mrkm4ntr) [reply pkg/volume/util/util.go]: No, we can't.  `volume.EmptyDir.Medium` is `v1.StorageMediumDefault`, `v1.StorageMediumMemory` or `v1.StorageMediumHugePages`.  `v1.StorageMediumHugePages` was added recently and `volume.EmptyDir.Medium` became ternary from binary. But `!=` in `IsLocalEphemeralVolume` and `==` in `isVolumeEphemeral` were not changed then. I think only `v1.StorageMediumDefault` is ephemeral, so I think the original code has a bug.

File: pkg/kubelet/server/stats/volume_stat_calculator.go (hunk 2 above)
Comment (mrkm4ntr) [reply pkg/kubelet/server/stats/volume_stat_calculator.go]: I think it is just a naming distinction, but I'm not confident.

**Comment by mrkm4ntr** [comment]
@mattjmcnaughton Thank you for your help! I will fix build and tests.

File: pkg/volume/util/util.go (hunk 3 above)
Comment (mrkm4ntr) [reply pkg/volume/util/util.go]: I found that `volume.DownwardAPI` is mounted as tmpfs. It is not right treating volume.Downward API as local ephemeral volume, so I deleted this condition.
It is mentioned [here](https://github.com/kubernetes/kubernetes/pull/55447#discussion_r152144318).

File: pkg/volume/util/util.go (hunk 1 above)
Comment (mrkm4ntr) [reply pkg/volume/util/util.go]: According to [this](https://github.com/kubernetes/kubernetes/pull/55447#issue-151801859), HugePages is not a local ephemeral volume.

**Review by mattjmcnaughton — APPROVED** [review]
/lgtm

Lgtm... I'd appreciate a double check from a community member on some of the points we brought up in our convo above.

**Comment by mrkm4ntr** [comment]
@saad-ali @yujuhong Could you review this?

**Comment by yujuhong** [comment]
/assign @jingxu97 

for clarification on the definition of local ephemeral storage

File: pkg/kubelet/server/stats/volume_stat_calculator.go (hunk 2 above)
Comment (jingxu97) [reply pkg/kubelet/server/stats/volume_stat_calculator.go]: For persistent, we could have local or remote. But for ephemeral, we only have local ephemeral.

**Comment by mrkm4ntr** [comment]
@saad-ali @jingxu97 LGTMed this. Could you approve this?

File: pkg/kubelet/server/stats/volume_stat_calculator.go (hunk 2 above)
Comment (pohly) [reply pkg/kubelet/server/stats/volume_stat_calculator.go:127]: I disagree. A "CSI ephemeral inline" volume could also be remote, if that is how the CSI driver implements the volume. The upcoming "generic ephemeral inline volumes" will use a normal PVC for provisioning, and then the "local" part no longer applies at all.

 I would rename "IsLocalEphemeralVolume" to "IsEphemeralVolume".

Does it matter anywhere whether the volume is local or remote?

**Comment by pohly** [comment]
I had another look at how `IsLocalEphemeralVolume` is used and it looks like the expectation is indeed that the volume is "local" - whatever that means, because several comments use it without explaining that. The definition of `IsLocalEphemeralVolume` itself is circular because the description just repeats the function name without adding any additional clarification.

Sorry, I have to punt on reviewing this. It touches code that I am simply not familiar with.

**Comment by cofyc** [comment]
Is this ok to merge?

File: pkg/volume/util/util.go (hunk 1 above)
Comment (cofyc) [reply pkg/volume/util/util.go:548]: why Secret volume is not an ephemeral volume?

Comment (mrkm4ntr) [reply pkg/volume/util/util.go:548]: @cofyc Please refer this.
https://github.com/kubernetes/kubernetes/pull/82401#issuecomment-529743428

Comment (cofyc) [reply pkg/volume/util/util.go:548]: Thanks!

**Comment by pohly** [comment]
> Is this ok to merge?

Given all the confusion around why some volume types are reported in IsLocalEphemeralVolume and others are not, I think it would be good to clearly document in the comment for that function what a "local ephemeral volume" is (not by listing them, but by describing the common attributes that matter and that distinguish them from other, similar volumes).

**Comment by cofyc** [comment]
I agree. The “ephemeral“ and "local" terms are confusing here. Here is my understanding:

From the purpose perspective, this function is used to determine whether the storage usage of volume should be calculated in the pod [ephemeral storage resource](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#local-ephemeral-storage) usage. Current usage:

- [volumemanager](https://github.com/kubernetes/kubernetes/blob/e7ca64fbe16d0c4b6c7b36aecde9cd75042b2828/pkg/kubelet/volumemanager/cache/desired_state_of_world.go#L259-L269) uses it to determine if pod's `ephemeral-storage` resource limit should be used
- [eviction manager](https://github.com/kubernetes/kubernetes/blob/e7ca64fbe16d0c4b6c7b36aecde9cd75042b2828/pkg/kubelet/eviction/helpers.go#L415-L424) uses it to calculate local ephemeral storage usage and evict the pod if the usage exceeds the limit (with other policies)

The common attributes are:

- Ephemeral (pod lifecycle)
  - the life cycle of `hostPath` volume is the same as the node, so it's ignored
- Consume local disk managed [by the kubelet](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#resource-emphemeralstorage-consumption)
  - this is why the secret volume is ignored, as it's backed by a tmpfs and kubelet tracks it as container memory usage (same as `tmpfs/hugePage` emptyDir)
 
The "CSI Ephemeral Inline Volume" is tied to the pod, but the storage is not managed by the kubelet, so it does not match the criteria.

**Comment by mrkm4ntr** [comment]
@cofyc Thank you for the great summary!
So how I change this PR? Adding a comment to `IsLocalEphemeralVolume`?

**Comment by cofyc** [comment]
yes, we can document it clearly in the comment as @pohly suggested

File: pkg/volume/util/util.go [hunk 4]
```diff
@@ -581,10 +581,12 @@ func GetPluginMountDir(host volume.VolumeHost, name string) string {
 
 // IsLocalEphemeralVolume determines whether the argument is a local ephemeral
 // volume vs. some other type
+// Local means the volume is a local disk managed by kubelet
+// Ephemeral means the lifecycle of the volume is the same as the Pod
```
Comment (pohly) [inline pkg/volume/util/util.go:585]: nit: punctuation is missing after each of the three sentences.

I would say "Local means the volume is using storage from the local disk that is managed by kubelet.".

Comment (mrkm4ntr) [reply pkg/volume/util/util.go]: Thanks. Fixed.

**Comment by mrkm4ntr** [comment]
@saad-ali @yujuhong PTAL.

File: pkg/kubelet/server/stats/volume_stat_calculator.go (hunk 2 above)
Comment (pohly) [reply pkg/kubelet/server/stats/volume_stat_calculator.go:127]: We clarified this in the definition of IsLocalEphemeralVolume: that the volume is local (= managed by kubelet) does matter.

**Comment by mrkm4ntr** [comment]
/remove-lifecycle rotten
I think this is still valid.

**Comment by pohly** [comment]
/assign @msau42 

For approval.

**Comment by derekwaynecarr** [comment]
The kubelet changes look good to me.

/lgtm
/approve

**Comment by mrkm4ntr** [comment]
Resolved conflicts. Could someone give lgtm?

**Pull request merged**